src/gc/gc.cpp

```diff
@@ -37595,3 +37595,22 @@ void PopulateDacVars(GcDacVars *gcDacVars)
 #endif // MULTIPLE_HEAPS
 #endif // DACCESS_COMPILE
 }
+
+void GCHeap::InvokeMemoryLoadChangeNotification(float memory_percent)
```

Comment:
Where is the GC going to invoke this notification?

I think the GC will have hard time to invoke the notification. Instead, I think this should be invoked on finalizer thread. It can be either special cased or it can be done via finalizable object that will get reregistered for finalization after every GC.

Reply:
Ideally, this would replace the `Gen2GcCallback` that we have in CoreLib today.

Reply:
Agreed. I don't really think its appropriate to attempt to call from the GC codebase into managed in very many points at all. (For instance, if a server GC worker thread did it, I believe you'll cause some truly interesting failures.) I'd be happy to see it happen on the finalizer thread, or possibly finding a way to shovel the notification onto the threadpool (which would be great, but I think somewhat difficult to pull off in a reasonable way).

Reply:
Yes, that's an incomplete implementation and discussing the matter with @davidwrighton and @Maoni0 I am going to invoke that on the Finalizer thread.

Reply:
@Maoni0 
I want to get notified before GC starts to work. I need synchronous callback before GC, which guarantees that GC would not start to work before the callback returns. I know this might impact performance of GC. But if nobody registers such callback, there would be no impact to GC. If somebody registers such callback, he must be responsible for the impact, that means he is sure that the impact is nothing serious in his use case/scenario. Of course, the synchronous callback before GC must be short and fast, and can not make any heap allocation. See https://github.com/dotnet/coreclr/issues/16931

Reply:
@jkotas this function is now invoked on the Finalizer thread.

Reply:
The Gen2 GC has to run now before the callbacks get send. Is it the behavior we like? Would it be better to deliver these callbacks more eagerly, e.g. after each GC?

>  this would replace the `Gen2GcCallback` that we have in CoreLib today

I meant that the `Gen2GcCallback` should be deleted and the code that is using `Gen2GcCallback` today would be switched to use this API instead. `Gen2GcCallback` was a workaround for the lack of this API.